### DISCUSSION TITLE
Shipping a separate PHAR using newer dependencies for PHP 8

### DISCUSSION THREAD

**Discussion by llaville** [composer/composer#11430] [Show and tell]
Hello,

Following recommandation of @stof, I've splitted initial discussion https://github.com/composer/composer/discussions/11429 into 2 topics.

To demonstrate how we may reproduce the Symfony Console v6 component conflict, here are steps to follow : 

- Have a source code that uses the Symfony Console v6 component: https://github.com/symfony/console is a very best candidate !

```shell
git clone -b v6.2.8 https://github.com/symfony/console
```
Remember that all versions 6.x are impacted by this commit https://github.com/symfony/console/commit/fbe9fdcfd458d268caa122fd01ef1367cc017bcc

- Modify the `composer.json` to add the `post-update-cmd` script as follow 

```json
    "minimum-stability": "stable",
    "scripts": {
        "post-update-cmd": "Symfony\\Component\\Console\\ComposerCompatibility::verify"
    }
```
NB: `miminum-stability` was changed from dev to stable only to demonstrate that is affect all versions and not only the "unstable" dev version (this change is not mandatory).

Add at root dir of the clone source code the following file `ComposerCompatibility.php` with such contents : 

```php
<?php declare(strict_types=1);

namespace Symfony\Component\Console;

use Composer\Script\Event;
use Symfony\Component\Console\Input\ArrayInput;

class ComposerCompatibility
{
    public static function verify(Event $event)
    {
        $composer = $event->getComposer();

        $vendorDir = $composer->getConfig()->get('vendor-dir');
        require_once $vendorDir . '/autoload.php';

        $input = new ArrayInput([]);

        if ($input->hasParameterOption('--whatever')) {

        }
    }
}
```
And finally install dependencies as usual with `composer update` command.

You should see at least : 

```text
> Symfony\Component\Console\ComposerCompatibility::verify

Fatal error: Declaration of Symfony\Component\Console\Input\ArrayInput::hasParameterOption(array|string $values, bool $onlyParams = false): bool must be compatible with Symfony\Component\Console\Input\InputInterface::hasParameterOption($values, bool $onlyParams = false) in /shared/backups/github/console/Input/ArrayInput.php on line 50
```

**Comment by llaville** [comment]
What I 'll propose is to upgrade your [Release Workflow](https://github.com/composer/composer/blob/2.5.5/.github/workflows/release.yml) to build two versions : one for PHP 7 support and another one for PHP 8 support only.

**Comment by llaville** [comment]
**Contextual information**

<details>
<summary>dependencies installed with Composer 2.5.5</summary>

```text
  - Installing composer/ca-bundle (1.3.5): Extracting archive
  - Installing symfony/polyfill-php80 (v1.27.0): Extracting archive
  - Installing symfony/deprecation-contracts (v2.5.2): Extracting archive
  - Installing symfony/finder (v5.4.21): Extracting archive
  - Installing composer/pcre (2.1.0): Extracting archive
  - Installing composer/class-map-generator (1.0.0): Extracting archive
  - Installing composer/metadata-minifier (1.0.0): Extracting archive
  - Installing composer/semver (3.3.2): Extracting archive
  - Installing composer/spdx-licenses (1.5.7): Extracting archive
  - Installing psr/log (1.1.4): Extracting archive
  - Installing composer/xdebug-handler (3.0.3): Extracting archive
  - Installing justinrainbow/json-schema (5.2.12): Extracting archive
  - Installing phpstan/phpstan (1.10.7): Extracting archive
  - Installing phpstan/phpstan-deprecation-rules (1.1.3): Extracting archive
  - Installing phpstan/phpstan-phpunit (1.3.10): Extracting archive
  - Installing phpstan/phpstan-strict-rules (1.5.0): Extracting archive
  - Installing phpstan/phpstan-symfony (1.2.23): Extracting archive
  - Installing react/promise (v2.9.0): Extracting archive
  - Installing seld/jsonlint (1.9.0): Extracting archive
  - Installing seld/phar-utils (1.2.1): Extracting archive
  - Installing seld/signal-handler (2.0.1): Extracting archive
  - Installing symfony/polyfill-mbstring (v1.27.0): Extracting archive
  - Installing symfony/polyfill-intl-normalizer (v1.27.0): Extracting archive
  - Installing symfony/polyfill-intl-grapheme (v1.27.0): Extracting archive
  - Installing symfony/polyfill-ctype (v1.27.0): Extracting archive
  - Installing symfony/string (v5.4.21): Extracting archive
  - Installing psr/container (1.1.1): Extracting archive
  - Installing symfony/service-contracts (v2.5.2): Extracting archive
  - Installing symfony/polyfill-php73 (v1.27.0): Extracting archive
  - Installing symfony/console (v5.4.21): Extracting archive
  - Installing symfony/filesystem (v5.4.21): Extracting archive
  - Installing symfony/phpunit-bridge (v6.2.7): Extracting archive
  - Installing symfony/polyfill-php81 (v1.27.0): Extracting archive
  - Installing symfony/process (v5.4.21): Extracting archive
```

</details>

See job <https://github.com/composer/composer/actions/runs/4478193862/jobs/7870657238>

I know it's normal, because you 'll keep to support PHP 7 on Composer LTS.

**Comment by Seldaek** [comment]
I don't see why we need this right now? This is only a problem for people requiring composer/composer, which is a bad practice anyway.. What problem are you trying to resolve? If it's about the CycloneDX thing then as I said I think this should be done in a separate step during build.

**Comment by llaville** [comment]
I wanted to update my manifests assets (see https://github.com/composer/composer/discussions/11429#discussioncomment-5617196) by automation by running a hook on `post-update-cmd` event.
I'll solve my problem by a composer plugin that will do the job !

**Reply by stof** [reply]
you could still do it by running a script instead of running a PHP callable.

Composer plugins won't solve your problem if they expect to load incompatible code in the composer process. They get loaded in the composer process and so must be compatible.

**Reply by llaville** [reply]
> you could still do it by running a script instead of running a PHP callable.

👍 I agree, but it's not what I want here !


> Composer plugins won't solve your problem if they expect to load incompatible code in the composer process. They get loaded in the composer process and so must be compatible.

👎 I disagree here. I've begun to change format of `bartlett/box-manifest` to make it composer plugin compatible, and all the tests I've done until now, run as I'm expected.

**Reply by stof** [reply]
Well, if you wrote your composer plugin to be compatible with the dependencies bundled in the Phar, it would indeed work. But doing that can also be done when using a PHP callable rather than a plugin.
